File: src/main/java/al132/alchemistry/Alchemistry.java
package al132.alchemistry;

import al132.alchemistry.network.NetworkHandler;
import al132.alchemistry.setup.ClientProxy;
import al132.alchemistry.setup.IProxy;
import al132.alchemistry.setup.ServerProxy;
import al132.alchemistry.utils.IngredientStack;
import al132.alib.ModData;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufAllocator;
import io.netty.buffer.PooledByteBufAllocator;
import net.minecraft.block.Block;
import net.minecraft.inventory.container.ContainerType;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.Items;
import net.minecraft.item.crafting.IRecipeSerializer;
import net.minecraft.item.crafting.Ingredient;
import net.minecraft.network.PacketBuffer;
import net.minecraft.tileentity.TileEntityType;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.DeferredWorkQueue;
import net.minecraftforge.fml.DistExecutor;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.event.server.FMLServerAboutToStartEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.fml.loading.FMLPaths;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod("alchemistry")
public class Alchemistry {

    public static final Logger LOGGER = LogManager.getLogger();

    public static IProxy proxy = DistExecutor.runForDist(() -> () -> new ClientProxy(), () -> () -> new ServerProxy());

    public static ModData data = new AlchemistryData();

    public static final String MODID = "alchemistry";


    public Alchemistry() {
        //ModLoadingContext.get().registerConfig(ModConfig.Type.CLIENT, Config.CLIENT_CONFIG);
        ModLoadingContext.get().registerConfig(ModConfig.Type.COMMON, Config.COMMON_CONFIG);

        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::setup);

        MinecraftForge.EVENT_BUS.register(this);

        Config.loadConfig(Config.COMMON_CONFIG, FMLPaths.CONFIGDIR.get().resolve("alchemistry-common.toml"));
    }


    private void setup(final FMLCommonSetupEvent e) {
        proxy.init();
        DeferredWorkQueue.runLater(() -> {
            NetworkHandler.register();
        });
        //ModRecipes.init();
    }

    @Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD)
    public static class RegistryEvents {

        @SubscribeEvent
        public static void onTilesRegistry(final RegistryEvent.Register<TileEntityType<?>> e) {
            data.registerTiles(e);
        }

        @SubscribeEvent
        public static void onBlocksRegistry(final RegistryEvent.Register<Block> e) {
            Ref.initBlocks();
            data.BLOCKS.forEach(e.getRegistry()::register);
        }

        @SubscribeEvent
        public static void onContainerRegistry(final RegistryEvent.Register<ContainerType<?>> e) {
            data.registerContainers(e);
        }

        @SubscribeEvent
        public static void onItemsRegistry(final RegistryEvent.Register<Item> e) {
            Ref.initItems();
            data.ITEMS.forEach(e.getRegistry()::register);
            data.BLOCKS.forEach(x -> e.getRegistry()
                    .register(new BlockItem(x, new Item.Properties().group(data.itemGroup)).setRegistryName(x.getRegistryName())));
        }

        @SubscribeEvent
        public static void onRecipeSerializerRegistry(final RegistryEvent.Register<IRecipeSerializer<?>> e) {
            e.getRegistry().register(Ref.ATOMIZER_SERIALIZER.setRegistryName(MODID, "atomizer"));
            e.getRegistry().register(Ref.COMBINER_SERIALIZER.setRegistryName(MODID, "combiner"));
            e.getRegistry().register(Ref.DISSOLVER_SERIALIZER.setRegistryName(MODID, "dissolver"));
            e.getRegistry().register(Ref.EVAPORATOR_SERIALIZER.setRegistryName(MODID, "evaporator"));
            e.getRegistry().register(Ref.FISSION_SERIALIZER.setRegistryName(MODID, "fission"));
            e.getRegistry().register(Ref.LIQUIFIER_SERIALIZER.setRegistryName(MODID, "liquifier"));
        }
    }
}

File: src/main/kotlin/al132/alchemistry/Alchemistry.kt
package al132.alchemistry

import al132.alchemistry.blocks.ModBlocks
import al132.alchemistry.items.ModItems
import crafttweaker.CraftTweakerAPI
import crafttweaker.IAction
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.SidedProxy
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLLoadCompleteEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.apache.logging.log4j.Logger
import java.util.*


//TODO: Everything
@Mod(modid = Reference.MODID,
        name = Reference.MODNAME,
        version = Reference.VERSION,
        dependencies = Reference.DEPENDENCIES,
        modLanguageAdapter = "net.shadowfacts.forgelin.KotlinAdapter")
object Alchemistry {

    //https://github.com/jaredlll08/ModTweaker/blob/1.12/src/main/java/com/blamejared/ModTweaker.java
    val LATE_REMOVALS: LinkedList<IAction> = LinkedList()
    val LATE_ADDITIONS: LinkedList<IAction> = LinkedList()

    lateinit var logger: Logger

    @SidedProxy(clientSide = "al132.alchemistry.ClientProxy", serverSide = "al132.alchemistry.CommonProxy")
    var proxy: CommonProxy? = null

    @EventHandler
    fun preInit(e: FMLPreInitializationEvent) = proxy!!.preInit(e)

    @EventHandler
    fun init(e: FMLInitializationEvent) = proxy!!.init(e)

    @EventHandler
    fun postInit(e: FMLPostInitializationEvent) = proxy!!.postInit(e)


    @EventHandler
    fun loadComplete(e: FMLLoadCompleteEvent) {
        try {
            LATE_REMOVALS.forEach(CraftTweakerAPI::apply)
            LATE_ADDITIONS.forEach(CraftTweakerAPI::apply)
        } catch (e: Exception) {
            e.printStackTrace()
            CraftTweakerAPI.logError("Error while applying actions", e)
        }
        LATE_REMOVALS.clear()
        LATE_ADDITIONS.clear()
    }

    @Mod.EventBusSubscriber(modid = Reference.MODID)
    object Registration {
        @JvmStatic
        @SubscribeEvent
        fun registerBlocks(event: RegistryEvent.Register<Block>) {
            ModBlocks.registerBlocks(event)
        }

        @JvmStatic
        @SubscribeEvent
        fun registerItems(event: RegistryEvent.Register<Item>) {
            ModBlocks.registerItemBlocks(event)
            ModItems.registerItems(event)
        }

        @SideOnly(Side.CLIENT)
        @JvmStatic
        @SubscribeEvent
        fun registerModels(event: ModelRegistryEvent) {
            ModBlocks.registerModels()
            ModItems.registerModels()
        }
    }
}